worker.go

```diff
@@ -3,373 +3,204 @@ package main
+	// TODO(sean) If we see the need to support various clean up strategies,
+	// we should just make this step plugable. For example, maybe instead of
+	// deleting, we want to move files to a done directory.
+	if w.DeleteFilesOnSuccess {
+		// Clean up data, meta and done files.
+		for _, name := range []string{dataPath, metaPath, donePath} {
+			if err := os.Remove(name); err != nil {
+				return fmt.Errorf("failed to clean up %s", name)
 			}
 		}
 
-		if !PretendUpload {
-			s3key := filepath.Join(s3path, targetNameData)
-
-			// check and skip if file exists
-			input := &s3.ListObjectsV2Input{
-				Bucket:  aws.String(s3bucket),
-				Prefix:  aws.String(s3key),
-				MaxKeys: aws.Int64(2),
-			}
-
-			var result *s3.ListObjectsV2Output
-			result, err = svc.ListObjectsV2(input)
-			if err != nil {
-				if aerr, ok := err.(awserr.Error); ok {
-					switch aerr.Code() {
-					case s3.ErrCodeNoSuchBucket:
-						fmt.Println(s3.ErrCodeNoSuchBucket, aerr.Error())
-					default:
-						fmt.Println(aerr.Error())
-					}
-				} else {
-					// Print the error, cast err to awserr.Error to get the Code and
-					// Message from an error.
-					fmt.Println(err.Error())
-				}
-				err = nil
-			} else {
-
-				//fmt.Println(result)
-
-				if len(result.Contents) == 2 {
-					fmt.Println("Files already exist in S3")
-					skipUpload = true
-				}
+		// Attempt to clean up parent directories up to root/node-xyz/uploads.
+		//
+		// NOTE(sean) There is a possible race condition with the upload agent here.
+		//
+		// It's possible that the upload agent creates the parent paths which are removed
+		// before we can upload. In this case, that particular rsync will fail and then
+		// will be tried again later.
+		//
+		// In order for this to happen, the OSN loader would have to upload and clean up the
+		// last staged item for a task right when that task is posting a new upload. This seems
+		// potentially rare enough that I'd opt for simpler, more robust cleanup logic for now.
+		for p := filepath.Dir(dataPath); filepath.Base(p) != "uploads"; p = filepath.Dir(p) {
+			if err := os.Remove(p); err != nil {
+				break
 			}
 		}
 	}
 
-	if !skipUpload {
-		var sageFileUrl string
-		sageFileUrl, err = uploadFile(uploadTarget, s3path, dataFileLocal, targetNameData, s3metadata)
-		if err != nil {
-			return
-		}
-
-		_, err = uploadFile(uploadTarget, s3path, metaFileLocal, targetNameMeta, nil)
-		if err != nil {
-			return
-		}
+	return nil
+}
 
-		fmt.Printf("upload success: %s %s (and .meta)\n", s3path, targetNameData)
+func readMetaFile(name string, m *MetaData) error {
+	var data struct {
+		EpochNano    *int64            `json:"ts"`
+		EpochNanoOld *int64            `json:"timestamp"` // only for reading (deprecated soon), keep it backwards compatible
+		Shasum       *string           `json:"shasum"`
+		Meta         map[string]string `json:"meta"`
+		MetaOld      map[string]string `json:"labels"` // only read (will write to meta)
```

Comment:
How long we gotta keep this around? I assume this is populated with `pywaggle` and it is using the new `meta`.

I would assume there are no applications still using the old `labels` ?